### PR TITLE
Move k8s client properties to record

### PR DESCRIPTION
**Pull request by wind57**

### CODE REVIEW

**Comment by wind57** [comment]
dependent on [this issue](https://github.com/spring-cloud/spring-cloud-commons/issues/1158) and [this PR](https://github.com/spring-cloud/spring-cloud-kubernetes/pull/1117)

File: spring-cloud-kubernetes-fabric8-autoconfig/src/test/java/org/springframework/cloud/kubernetes/fabric8/Fabric8ClientUserAgentEnvPropertyTests.java [hunk 1]
```diff
@@ -34,6 +35,7 @@
  * test "User-Agent" functionality via system properties
  */
 @SpringBootTest(classes = App.class, properties = "spring.main.cloud-platform=KUBERNETES")
+@DirtiesContext
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-autoconfig/src/test/java/org/springframework/cloud/kubernetes/fabric8/Fabric8ClientUserAgentEnvPropertyTests.java:38]: so, the unit test kept failing for me and this is the only way I made it work.

File: spring-cloud-kubernetes-fabric8-autoconfig/src/main/java/org/springframework/cloud/kubernetes/fabric8/Fabric8AutoConfiguration.java [hunk 2]
```diff
@@ -22,8 +22,6 @@
 import io.fabric8.kubernetes.client.ConfigBuilder;
 import io.fabric8.kubernetes.client.DefaultKubernetesClient;
 import io.fabric8.kubernetes.client.KubernetesClient;
-import org.apache.commons.logging.Log;
```
Comment (wind57) [inline spring-cloud-kubernetes-fabric8-autoconfig/src/main/java/org/springframework/cloud/kubernetes/fabric8/Fabric8AutoConfiguration.java:25]: just renames and small clean-ups

File: spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/KubernetesClientProperties.java [hunk 3]
```diff
@@ -28,7 +29,14 @@
  * @author Ioannis Canellos
  */
 @ConfigurationProperties(PREFIX)
-public class KubernetesClientProperties {
+public record KubernetesClientProperties(Boolean trustCerts, String masterUrl, String apiVersion, String namespace,
```
Comment (wind57) [inline spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/KubernetesClientProperties.java:32]: move this one to a record is the main idea of this PR

File: spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/KubernetesConfigDataLocationResolver.java [hunk 4]
```diff
@@ -170,25 +169,25 @@ private static PropertyHolder of(ConfigDataLocationResolverContext context) {
 			String namespace = binder.bind("spring.cloud.kubernetes.client.namespace", String.class)
 					.orElse(binder.bind("kubernetes.namespace", String.class).orElse(""));
 
-			KubernetesClientProperties kubernetesClientProperties = clientProperties(context);
```
Comment (wind57) [inline spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/KubernetesConfigDataLocationResolver.java:173]: this is a little tricky. What we used to do, is create a `KubernetesClientProperties` and then change its namespace. I do the same thing here, but via a copy constructor in `KubernetesClientProperties` called `withNamespace`, because that is now a record.

**Comment by wind57** [comment]
@ryanjbaxter I need a bit of your help please :( integration tests are failing with a weird exception logs:

```
	  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::            (v3.0.0-RC1)

2022-10-27T17:51:02.135Z  INFO 1 --- [           main] o.s.c.k.d.DiscoveryServerApplication     : Starting DiscoveryServerApplication v3.0.0-SNAPSHOT using Java 17.0.5 on spring-cloud-kubernetes-discoveryserver-deployment-b46457fvdwnf with PID 1
 (/app.jar started by root in /)
2022-10-27T17:51:02.137Z  INFO 1 --- [           main] o.s.c.k.d.DiscoveryServerApplication     : The following 1 profile is active: "kubernetes"
2022-10-27T17:51:02.693Z  WARN 1 --- [           main] onfigReactiveWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException:
 Error creating bean with name 'discoveryInformerConfigurer' defined in class path resource [org/springframework/cloud/kubernetes/client/discovery/KubernetesDiscoveryClientAutoConfiguration$KubernetesInformerDiscoveryConfiguration.class]: Unsati
sfied dependency expressed through method 'discoveryInformerConfigurer' parameter 1: Error creating bean with name 'apiClient' defined in class path resource [org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.class]:
Unsatisfied dependency expressed through method 'apiClient' parameter 0: Error creating bean with name 'org.springframework.boot.context.internalConfigurationPropertiesBinder': Instantiation of supplied bean failed
2022-10-27T17:51:02.731Z ERROR 1 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'discoveryInformerConfigurer' defined in class path resource [org/springframework/cloud/kubernetes/client/discovery/KubernetesDiscoveryClientAutoConf
iguration$KubernetesInformerDiscoveryConfiguration.class]: Unsatisfied dependency expressed through method 'discoveryInformerConfigurer' parameter 1: Error creating bean with name 'apiClient' defined in class path resource [org/springframework/c
loud/kubernetes/client/KubernetesClientAutoConfiguration.class]: Unsatisfied dependency expressed through method 'apiClient' parameter 0: Error creating bean with name 'org.springframework.boot.context.internalConfigurationPropertiesBinder': Ins
tantiation of supplied bean failed
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:793) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:543) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1324) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1161) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:561) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:521) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:205) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:122) ~[spring-context-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:745) ~[spring-context-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:565) ~[spring-context-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:66) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:730) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:432) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:308) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:150) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframewok.cloud.kubernetes.discoveryserver.DiscoveryServerApplication.main(DiscoveryServerApplication.java:29) ~[classes!/:3.0.0-SNAPSHOT]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49) ~[app.jar:3.0.0-SNAPSHOT]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:95) ~[app.jar:3.0.0-SNAPSHOT]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:58) ~[app.jar:3.0.0-SNAPSHOT]
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:65) ~[app.jar:3.0.0-SNAPSHOT]
Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'apiClient' defined in class path resource [org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.class]: Unsatisf
ied dependency expressed through method 'apiClient' parameter 0: Error creating bean with name 'org.springframework.boot.context.internalConfigurationPropertiesBinder': Instantiation of supplied bean failed
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:543) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1324) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1161) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:561) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:521) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1375) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1295) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:880) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:784) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	... 26 common frames omitted
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.springframework.boot.context.internalConfigurationPropertiesBinder': Instantiation of supplied bean failed
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainInstanceFromSupplier(AbstractAutowireCapableBeanFactory.java:1236) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainFromSupplier(AbstractAutowireCapableBeanFactory.java:1210) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1157) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:561) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:521) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:205) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.boot.context.properties.ConfigurationPropertiesBinder.get(ConfigurationPropertiesBinder.java:210) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.context.properties.ConstructorBound.from(ConstructorBound.java:41) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.context.properties.ConfigurationPropertiesBeanRegistrar.lambda$createBeanDefinition$1(ConfigurationPropertiesBeanRegistrar.java:96) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainInstanceFromSupplier(AbstractAutowireCapableBeanFactory.java:1230) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainFromSupplier(AbstractAutowireCapableBeanFactory.java:1210) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1157) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:561) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:521) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.config.DependencyDescriptor.resolveCandidate(DependencyDescriptor.java:254) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1375) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1295) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:880) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:784) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	... 40 common frames omitted
Caused by: java.lang.IllegalStateException: PropertySources have not yet been applied
	at org.springframework.util.Assert.state(Assert.java:76) ~[spring-core-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.context.support.PropertySourcesPlaceholderConfigurer.getAppliedPropertySources(PropertySourcesPlaceholderConfigurer.java:223) ~[spring-context-6.0.0-RC2.jar!/:6.0.0-RC2]
	at org.springframework.boot.context.properties.PropertySourcesDeducer.getPropertySources(PropertySourcesDeducer.java:50) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.context.properties.ConfigurationPropertiesBinder.<init>(ConfigurationPropertiesBinder.java:85) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.context.properties.ConfigurationPropertiesBinder$Factory.create(ConfigurationPropertiesBinder.java:229) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.boot.context.properties.ConfigurationPropertiesBinder.lambda$register$0(ConfigurationPropertiesBinder.java:202) ~[spring-boot-3.0.0-RC1.jar!/:3.0.0-RC1]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.obtainInstanceFromSupplier(AbstractAutowireCapableBeanFactory.java:1230) ~[spring-beans-6.0.0-RC2.jar!/:6.0.0-RC2]
	... 65 common frames omitted
```

So it says that it can't create the `ApiClient` in : `org.springframework.cloud.kubernetes.client.KubernetesClientAutoConfiguration`

```
	@Bean
	@ConditionalOnMissingBean
	public ApiClient apiClient(KubernetesClientProperties properties) {
		ApiClient apiClient = kubernetesApiClient();
		apiClient.setUserAgent(properties.userAgent());
		return apiClient;
	}
```

can not be created because "PropertySources have not yet been applied", I assume the ones that came as arguments to `apiClient`. I am still trying to understand what is going on, but may be this is already known in some other projects that might face this?

**Comment by ryanjbaxter** [comment]
Which integration test is this?

**Comment by wind57** [comment]
Kuberneneste discovery it

**Comment by wind57** [comment]
this is what happens when you respond at 2 AM, sorry! 

So the test is `DiscoveryClientIT`, but you will not see the logs in the build because these logs come from the deployment pod of `spring-cloud-kubernetes-discoveryserver`

**Comment by wind57** [comment]
so this is indeed something to do with records. If I change `KubernetesClientProperties` to a class, and have it something like this:

```
@ConfigurationProperties(PREFIX)
public class KubernetesClientProperties {

	private Boolean trustCerts;
        // ... other fields

        public Boolean trustCerts() {
		return trustCerts;
	}

        public void setTrustCerts(Boolean trustCerts) {
		this.trustCerts = trustCerts;
	}
```

that is: I provide setters, instead of constructor, then it works. Not sure if it helps, but I am still trying to figure out what is happening.

**Comment by wind57** [comment]
it's actually easier to reproduce, just checkout this branch, build it, and run `DiscoveryServerApplication`

**Comment by wind57** [comment]
OK, so I am onto something here. It seems that when I moved `KubernetesClientProperties` to a record, I only expose a problem that we already had. I don't yet fully understand what is going on, but if I do: 

- switch to `main` method
- add two env variables `spring.main.cloud-platform=KUBERNETES` and `spring.cloud.kubernetes.client.user-agent=agent007`
- change a Bean that we already have to:

```
	@Bean
	@ConditionalOnMissingBean
	public ApiClient apiClient(KubernetesClientProperties properties, Environment environment) {
		ApiClient apiClient = kubernetesApiClient();
		apiClient.setUserAgent(properties.getUserAgent());

		System.out.println("properties.agent = " + properties.getUserAgent());
		System.out.println("evn.agent = " + environment.getProperty(KubernetesClientProperties.PREFIX+".userAgent"));
		return apiClient;
	}
```

basically add `Environment` and print two things.

- then run `DiscoveryServerApplication`, to my surprise, the output is:

```
properties.agent = Spring-Cloud-Kubernetes-Application
evn.agent = agent007
```

So properties are not bind to `KubernetesClientProperties`? hmm

File: spring-cloud-kubernetes-client-autoconfig/src/main/java/org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.java [hunk 5]
```diff
@@ -42,9 +42,9 @@ public class KubernetesClientAutoConfiguration {
 
 	@Bean
 	@ConditionalOnMissingBean
-	public ApiClient apiClient(KubernetesClientProperties properties) {
```
Comment (wind57) [inline spring-cloud-kubernetes-client-autoconfig/src/main/java/org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.java:45]: @ryanjbaxter I am in the blur. May be I spent too much time on this and my vision is somehow biased, but I don't understand what is going on. If we inject `KubernetesClientProperties` in here, properties are not injected. 

You left somehow the same comment in `KubernetesInformerDiscoveryConfiguration`:

```// Injecting KubernetesDiscoveryProperties here would cause it to be
			// initialize too early
			// Instead get the all-namespaces property value from the Environment directly
			boolean allNamespaces = environment.getProperty("spring.cloud.kubernetes.discovery.all-namespaces",
					Boolean.class, false);
```

but for the love of God, I do not understand why that happens.

Comment (wind57) [reply spring-cloud-kubernetes-client-autoconfig/src/main/java/org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.java:45]: We can move this to main, since it moves properties to a record and then l'll try to figure out what is going on. Either way, this never actually worked the way we thought.

Comment (ryanjbaxter) [reply spring-cloud-kubernetes-client-autoconfig/src/main/java/org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.java:45]: Yeah, I've run into this several times.  Its all about the order in which things get created.  If something gets created too early before the environment is populated then the resulting configuration properties bean won't be initialized properly.  By using the environment we are basically getting the property "on demand".  I am curious though do you know what change is causing the initialization so early?

Comment (wind57) [reply spring-cloud-kubernetes-client-autoconfig/src/main/java/org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.java:45]: Here is the thing: no change at all. This was never working. I debugged in main branch, and properties are initialized with all "defaults". So when we had a default constructor, that was called. Now that we don't ( because of record ), things break. 

I'm a bit busy with my _actual_ work this week :),  but I'll try to figure out end to end what the problem is

Comment (ryanjbaxter) [reply spring-cloud-kubernetes-client-autoconfig/src/main/java/org/springframework/cloud/kubernetes/client/KubernetesClientAutoConfiguration.java:45]: Got it can you create an issue for it and if I have some cycles I can try and take a look as well.  Appreciate all your help!
